tests/robustness/validate/patch_history.go

```diff
@@ -54,28 +57,36 @@ func uniqueWatchEvents(reports []report.ClientReport) map[model.Event]traffic.Ti
 	return persisted
 }
 
-func patchOperationsWithWatchEvents(operations []porcupine.Operation, watchEvents map[model.Event]traffic.TimedWatchEvent) []porcupine.Operation {
-
+func patchOperations(operations []porcupine.Operation, watchEvents map[model.Event]traffic.TimedWatchEvent, persistedOperations map[model.EtcdOperation]int64) []porcupine.Operation {
 	newOperations := make([]porcupine.Operation, 0, len(operations))
 	lastObservedOperation := lastOperationObservedInWatch(operations, watchEvents)
 
 	for _, op := range operations {
 		request := op.Input.(model.EtcdRequest)
 		resp := op.Output.(model.MaybeEtcdResponse)
-		if resp.Error == "" || op.Call > lastObservedOperation.Call || request.Type != model.Txn {
+		if resp.Error == "" || request.Type != model.Txn {
 			// Cannot patch those requests.
 			newOperations = append(newOperations, op)
 			continue
 		}
-		event := matchWatchEvent(request.Txn, watchEvents)
-		if event != nil {
+		canMatchEvents := op.Call <= lastObservedOperation.Call
+		matchingEvent := matchWatchEvent(request.Txn, watchEvents)
+		if canMatchEvents && matchingEvent != nil {
+			eventTime := matchingEvent.Time.Nanoseconds()
 			// Set revision and time based on watchEvent.
-			op.Return = event.Time.Nanoseconds()
-			op.Output = model.MaybeEtcdResponse{PartialResponse: true, EtcdResponse: model.EtcdResponse{Revision: event.Revision}}
-			newOperations = append(newOperations, op)
-			continue
+			if eventTime < op.Return {
+				op.Return = eventTime
+			}
+			op.Output = model.MaybeEtcdResponse{PartialResponse: true, EtcdResponse: model.EtcdResponse{Revision: matchingEvent.Revision}}
 		}
-		if !canBeDiscarded(request.Txn) {
+		persistedReturnTime := matchReturnTime(request, persistedOperations)
+		if persistedReturnTime != nil {
+			// Set return time based on persisted return time.
+			if *persistedReturnTime < op.Return {
+				op.Return = *persistedReturnTime
+			}
+		}
+		if persistedReturnTime != nil || (len(persistedOperations) == 0 && canMatchEvents && matchingEvent != nil) || !canBeDiscarded(request.Txn) {
```

Comment:
panic if a watch event cannot be found in the persistedOperations (`len(persistedOperations) != 0 && persistedReturnTime == nil && canMatchEvents && matchingEvent != nil`?

Reply:
That's not correct, TXN request can be persisted in WAL, but the condition in TXN can fail, so operationsOnSuccess will not be executed, no watch events will be emitted.

Reply:
but can a watch event be emitted when the request is not persisted in WAL?

Reply:
No, but this is will be cough in watch validation.